### 본 코드는 HR data를 통해 다양한 예측을 하는 코드입니다. (전처리 편)
#### [Data Reference: Kaggle](https://www.kaggle.com/rhuebner/human-resources-data-set/data)

In [16]:
# Suengjae Hong (2018-06-12)
# Predicting something by using HR dataset
# I directly changed variable name in the fifth dataset (salary_grid)

In [252]:
library(dplyr)

setwd("~/kaggle/human-resources-data-set")
options(warn=-1)

dbList <- list.files(getwd(),pattern='*csv')
dbList

[1] "HRDataset_v9.csv"     "production_staff.csv" "recruiting_costs.csv"
[4] "salary_grid.csv"

본 데이터는 총 4개의 데이터로 구성이 되어 있습니다. 

HRDataset_v9는 전체적인 data set이고, 나머지는 부수적으로 확인을 하는 자료로 생각하면 될 것 같습니다.

In [253]:
hrdta <- list()

for(i in 1:length(dbList)){
    hrdta[[i]] <- read.csv(dbList[i],header=TRUE,fileEncoding="UTF-8")
    }

head(hrdta[[4]]) # to check each data set, change [[i]]

Position                     salary_min salary_mid salary_max hourly_min
1 Administrative Assistant     30000      40000      50000      14.42     
2 Sr. Administrative Assistant 35000      45000      55000      16.83     
3 Accountant I                 42274      51425      62299      20.32     
4 Accountant II                50490      62158      74658      24.27     
5 Sr. Accountant               63264      76988      92454      30.42     
6 Network Engineer             50845      66850      88279      24.44     
  hourly_mid hourly_max
1 19.23      24.04     
2 21.63      26.44     
3 24.72      29.95     
4 29.88      35.89     
5 37.01      44.45     
6 32.14      42.44

일단 모든 dataset을 하나의 리스트(list)로 담았습니다. 

간략하게 hrdta라는 리스트에는
(1) HRDataset_v9
(2) production_staff
(3) recruiting_costs
(4) salary_grid
의 데이터가 모두 하나의 리스트 변수에 담겨 있고, 이들을 확인하기 위해서는 hrdta[[n]]로써 해당 위치의 n을 입력하면 확인 할 수 있습니다. 

그 예시로써 hrdta[[4]]를 입력하여 salary_grid가 어떠한 데이터인지 확인하였습니다.

In [260]:
str(hrdta[[1]])

'data.frame':	310 obs. of  28 variables:
 $ Employee.Name      : Factor w/ 310 levels "Adinolfi, Wilson  K",..: 29 155 270 128 256 261 160 226 25 90 ...
 $ Employee.Number    : int  1103024456 1106026572 1302053333 1211050782 1307059817 711007713 1102024115 1206043417 1307060188 1201031308 ...
 $ MarriedID          : int  1 0 0 1 0 1 1 1 1 1 ...
 $ MaritalStatusID    : int  1 2 0 1 0 1 1 1 1 1 ...
 $ GenderID           : int  0 1 1 0 0 0 1 1 0 0 ...
 $ EmpStatus_ID       : int  1 1 1 1 1 5 1 5 1 1 ...
 $ DeptID             : int  1 1 1 1 1 1 1 1 1 1 ...
 $ Perf_ScoreID       : int  3 3 3 9 9 3 3 3 0 3 ...
 $ Age                : int  30 34 31 32 30 30 33 33 31 39 ...
 $ Pay.Rate           : num  28.5 23 29 21.5 16.6 ...
 $ State              : Factor w/ 28 levels "AL","AZ","CA",..: 11 11 11 11 11 11 11 11 11 11 ...
 $ Zip                : int  1450 1460 2703 2170 2330 1844 1460 2045 2468 2050 ...
 $ DOB                : Factor w/ 306 levels "1/12/1973","1/12/1974",..: 54 148 274 282 17

먼저 core data set인 HRDataset_v9 (이하 편의 상 hrdta[[1]] 등으로 칭하도록 하겠습니다)의 데이터 요약을 하였습니다. 

가장 먼저 눈에 띄는 건 Employee.Number (종업원의 id를 의미)가 integer로 표현이 되어 있다는 것입니다. 명목변수(id)이다보니 해당 변수를 factor로 변경시킬 예정입니다. (그 외에도 MarriedID, MartialStatusID, GenderID, EmpStatus_Id, DeptID 모두 명목 변수처럼 보입니다. 왜냐하면 MarriedID - 기혼 여부, MartialStatusID - 기혼 여부2, GenderID - 성별, DeptID - 부서 이기 때문입니다. 그러나 본 데이터에서는 description을 위하여 integer variable을 별도로 생성한 것 같습니다. 각각의 변수들은 Gender ID를 제외하고 MartialDesc, Employment.Status, Department을 numeric화 시켰습니다. 

따라서 해당 integer화 되어 있는 factor 변수들은 체크만 해두도록 하겠습니다. Performance ID의 경우는 예상으로는 평가자의 순서척도로 의해 결정되었기 때문에 이 역시 명목 변수처럼 보이지만, 조금 있다가 확인해보도록 하겠습니다.

In [255]:
# hrdta[[1]]$MarriedID <- as.factor(hrdta[[1]]$MarriedID)
# hrdta[[1]]$MaritalStatusID <- as.factor(hrdta[[1]]$MaritalStatusID)
# hrdta[[1]]$GenderID <- as.factor(hrdta[[1]]$GenderID)
# hrdta[[1]]$EmpStatus_ID <- as.factor(hrdta[[1]]$EmpStatus_ID)
# hrdta[[1]]$DeptID <- as.factor(hrdta[[1]]$DeptID)

"#"을 적을 경우 R에서는 주석 표시를 남길 수 있습니다.

In [256]:
str(hrdta[[1]][,2:7])

'data.frame':	310 obs. of  6 variables:
 $ Employee.Number: int  1103024456 1106026572 1302053333 1211050782 1307059817 711007713 1102024115 1206043417 1307060188 1201031308 ...
 $ MarriedID      : int  1 0 0 1 0 1 1 1 1 1 ...
 $ MaritalStatusID: int  1 2 0 1 0 1 1 1 1 1 ...
 $ GenderID       : int  0 1 1 0 0 0 1 1 0 0 ...
 $ EmpStatus_ID   : int  1 1 1 1 1 5 1 5 1 1 ...
 $ DeptID         : int  1 1 1 1 1 1 1 1 1 1 ...


확인하면, Employee.Number가 309개의 level로 되어 있다는 점을 확인할 수 있네요. 이는 전체 310개의 record에서 Employee.Number가 중복이 있다는 점을 확인 할 수 있습니다. 

따라서 해당 중복 데이터가 (1)두 개의 레코드가 들어간 인물인지 혹은 (2) 다른 사람인데 Id가 중복된 것인지를 확인 할 필요가 있습니다. 예상하면, (2)일 것입니다. 왜냐하면 Employee.Name은 310개의 level을 갖고 있기 때문입니다. 물론 이름을 바꾸었을 경우도 배제 할 수 없으니 확인해보도록 하겠습니다.

In [257]:
hrdta[[1]][which(duplicated(hrdta[[1]]$Employee.Number)),]

Employee.Name Employee.Number MarriedID MaritalStatusID GenderID
154 Ndzi, Horia   1204033041      1         1               1       
    EmpStatus_ID DeptID Perf_ScoreID Age Pay.Rate ... Date.of.Hire
154 5            5      3            35  22       ... 4/1/2013    
    Days.Employed Date.of.Termination Reason.For.Term Employment.Status     
154 1150          5/25/2016           more money      Voluntarily Terminated
    Department        Position                Manager.Name Employee.Source  
154 Production        Production Technician I Amy Dunn     Employee Referral
    Performance.Score
154 Fully Meets

현재 중복되었다고 판단되는 Employee를 반환하였습니다. 

그렇다면 1204033041 이라는 Employee.Number를 가진 사람이 한 명 더 있다는 것을 의미합니다. 그게 누구인지 확인하도록 하겠습니다.

In [258]:
check <- subset(hrdta[[1]],
               Employee.Number == 1204033041)
check

Employee.Name  Employee.Number MarriedID MaritalStatusID GenderID
153 Ndzi, Colombui 1204033041      0         0               1       
154 Ndzi, Horia    1204033041      1         1               1       
    EmpStatus_ID DeptID Perf_ScoreID Age Pay.Rate ... Date.of.Hire
153 5            5      3            29  18       ... 9/26/2011   
154 5            5      3            35  22       ... 4/1/2013    
    Days.Employed Date.of.Termination Reason.For.Term  Employment.Status     
153  921          4/4/2014            return to school Voluntarily Terminated
154 1150          5/25/2016           more money       Voluntarily Terminated
    Department        Position                Manager.Name   Employee.Source   
153 Production        Production Technician I Webster Butler Diversity Job Fair
154 Production        Production Technician I Amy Dunn       Employee Referral 
    Performance.Score
153 Fully Meets      
154 Fully Meets

확인하면 Ndzi, Colombui와 Ndzi, Horia가 중복된 Number를 갖고 있습니다. 이들에 대한 정보는 사실상 다른 정보이기 때문에 서로 다른 ID를 부여해야 할 것 같습니다. 

In [264]:
rm(check) # delete check variable 

hrdta1 <- hrdta[[1]]

hrdta1$Employee.Number <- replace(hrdta1$Employee.Number,
                                     hrdta1$Employee.Name == "Ndzi, Colombui",
                                     9999999999)

hrdta1$Employee.Number <- as.factor(hrdta1$Employee.Number)

nrow(hrdta1[which(duplicated(hrdta1$Employee.Number)),])

[1] 0

따라서 Horia Ndzi의 Employee ID를 편의상 9999999999 으로 바꾸었습니다.. 

이후 중복된 레코드의 수를 세어보니 0으로 나옵니다.

그리고 정제된 hrdta[[1]]을 hrdta1라는 변수로 만들었습니다. 

### 이 내용을 작성하며 알게 된 사실인데, 미리 as.factor로 변경을 할 경우 데이터 세트 상에 factor로 존재하는 값들은 일종의 집합을 형성하게 됩니다. 따라서 해당 factor 값과 같지 않은 다른 값이 들어오게 되면 해당 변수 값은 NA로 처리가 됩니다. 따라서 Employee.Number(ID)를 integer로 유지하다가 뒷 부분에서 factor (범주형) 변수로 변경을 하였습니다. 

### 또 다른 방식으로는 해당 ID를 string type으로 바꾸어서 할 수도 있을 것 같습니다.

In [265]:
str(hrdta[[2]]) # production staff

'data.frame':	256 obs. of  15 variables:
 $ Employee.Name      : Factor w/ 210 levels "","Adinolfi, Wilson  K",..: 5 22 28 50 79 88 93 113 130 152 ...
 $ Race.Desc          : Factor w/ 7 levels "","American Indian or Alaska Native",..: 7 3 7 7 7 7 3 7 5 7 ...
 $ Date.of.Hire       : Factor w/ 71 levels "","1/10/2011",..: 59 71 3 66 42 2 25 71 62 12 ...
 $ TermDate           : Factor w/ 77 levels "","1/11/2014",..: 1 63 1 1 1 18 69 1 1 42 ...
 $ Reason.for.Term    : Factor w/ 18 levels "","Another position",..: 12 16 12 12 12 3 18 12 12 2 ...
 $ Employment.Status  : Factor w/ 6 levels "","Active","Future Start",..: 2 6 2 2 2 5 6 2 2 6 ...
 $ Department         : Factor w/ 2 levels "","Production       ": 2 2 2 2 2 2 2 2 2 2 ...
 $ Position           : Factor w/ 4 levels "","Production Manager",..: 2 2 2 2 2 2 2 2 2 2 ...
 $ Pay                : Factor w/ 45 levels ""," $14.00 ",..: 44 39 45 40 43 38 37 45 42 36 ...
 $ Manager.Name       : Factor w/ 11 levels "","Amy Dunn",..: 6 6 6 6 6 

다음으로 2번째 dataset에 대한 요약을 해보았습니다. 총 256개의 변수 중 구분이 될만한 Employee.Name에는 210개가 있습니다. 이들 역시 중복된 데이터로 볼 수 있는데 단순 중복이 아니라 레코드가 누적됨에 따라 변화된 사람일수도 있습니다. 따라서 이들 정보에 대해 확인하도록 하겠습니다.

In [266]:
head(hrdta[[2]][which(duplicated(hrdta[[2]]$Employee.Name)),],10)

Employee.Name Race.Desc Date.of.Hire TermDate Reason.for.Term
211                                                              
212                                                              
213                                                              
214                                                              
215                                                              
216                                                              
217                                                              
218                                                              
219                                                              
220                                                              
    Employment.Status Department Position Pay Manager.Name Performance.Score
211                                                                         
212                                                                         
213                                                                         
214                                                                         
215                                                                         
216                                                                         
217                                                                         
218                                                                         
219                                                                         
220                                                                         
    Abutments.Hour.Wk.1 Abutments.Hour.Wk.2 Daily.Error.Rate X90.day.Complaints
211 NA                  NA                  NA               NA                
212 NA                  NA                  NA               NA                
213 NA                  NA                  NA               NA                
214 NA                  NA                  NA               NA                
215 NA                  NA                  NA               NA                
216 NA                  NA                  NA               NA                
217 NA                  NA                  NA               NA                
218 NA                  NA                  NA               NA                
219 NA                  NA                  NA               NA                
220 NA                  NA                  NA               NA

알고보니 NA값들이 입력되어 있는 불필요한 레코드입니다. 이들은 모두 삭제하도록 하겠습니다.

혹시라도 실제 값임에도 불구하고 Missing value가 들어가 있을수도 있으니 그러한 경우가 있는지 확인을 먼저 하도록 하겠습니다.

In [267]:
head(subset(hrdta[[2]],
      select=c(Employee.Name, 
               Abutments.Hour.Wk.1, 
               Abutments.Hour.Wk.2, 
               Daily.Error.Rate, 
               X90.day.Complaints),
      subset= is.na(hrdta[[2]]$Abutments.Hour.Wk.1)),10)

Employee.Name  Abutments.Hour.Wk.1 Abutments.Hour.Wk.2 Daily.Error.Rate
209 Woodson, Jason NA                  NA                  NA              
210                NA                  NA                  NA              
211                NA                  NA                  NA              
212                NA                  NA                  NA              
213                NA                  NA                  NA              
214                NA                  NA                  NA              
215                NA                  NA                  NA              
216                NA                  NA                  NA              
217                NA                  NA                  NA              
218                NA                  NA                  NA              
    X90.day.Complaints
209 NA                
210 NA                
211 NA                
212 NA                
213 NA                
214 NA                
215 NA                
216 NA                
217 NA                
218 NA

예상대로 확인해보니 Woodson, Jason의 경우는 이름이 있지만 Abutments.Hour.Wk.1 / 
Abutments.Hour.Wk.2 / 
Daily.Error.Rate / 
X90.day.Complaints 의 항목에서 missing value가 있습니다. (편의상 10개의 레코드만 보여드렸습니다. 전체적으로 확인해보아도 위 경우와 동일합니다)

함부로 지웠다가는 Woodson, Jason의 레코드를 날릴 뻔 했습니다. 따라서 이 사람을 제외하고 공란으로 들어 있는 레코드를 삭제하도록 하겠습니다.

In [268]:
hrdta2 <- subset(hrdta[[2]],hrdta[[2]]$Employee.Name != "" | !is.na(hrdta[[2]]$Abutments.Hour.Wk.1))
tail(hrdta2,3)

str(hrdta2)

Employee.Name      Race.Desc Date.of.Hire TermDate  Reason.for.Term     
207 Winthrop, Jordan   White     1/7/2013     2/21/2016 retiring            
208 Wolk, Hang  T      White     9/29/2014              N/A - still employed
209 Woodson, Jason     White     7/7/2014               N/A - still employed
    Employment.Status      Department        Position                 Pay     
207 Voluntarily Terminated Production        Production Technician II  $29.00 
208 Active                 Production        Production Technician II  $22.00 
209 Active                 Production        Production Technician II  $24.25 
    Manager.Name   Performance.Score Abutments.Hour.Wk.1 Abutments.Hour.Wk.2
207 Brannon Miller Exceeds           12                  11                 
208 David Stanley  Fully Meets       12                  17                 
209 Kissy Sullivan Fully Meets       NA                  NA                 
    Daily.Error.Rate X90.day.Complaints
207  0                0                
208  0                0                
209 NA               NA

'data.frame':	209 obs. of  15 variables:
 $ Employee.Name      : Factor w/ 210 levels "","Adinolfi, Wilson  K",..: 5 22 28 50 79 88 93 113 130 152 ...
 $ Race.Desc          : Factor w/ 7 levels "","American Indian or Alaska Native",..: 7 3 7 7 7 7 3 7 5 7 ...
 $ Date.of.Hire       : Factor w/ 71 levels "","1/10/2011",..: 59 71 3 66 42 2 25 71 62 12 ...
 $ TermDate           : Factor w/ 77 levels "","1/11/2014",..: 1 63 1 1 1 18 69 1 1 42 ...
 $ Reason.for.Term    : Factor w/ 18 levels "","Another position",..: 12 16 12 12 12 3 18 12 12 2 ...
 $ Employment.Status  : Factor w/ 6 levels "","Active","Future Start",..: 2 6 2 2 2 5 6 2 2 6 ...
 $ Department         : Factor w/ 2 levels "","Production       ": 2 2 2 2 2 2 2 2 2 2 ...
 $ Position           : Factor w/ 4 levels "","Production Manager",..: 2 2 2 2 2 2 2 2 2 2 ...
 $ Pay                : Factor w/ 45 levels ""," $14.00 ",..: 44 39 45 40 43 38 37 45 42 36 ...
 $ Manager.Name       : Factor w/ 11 levels "","Amy Dunn",..: 6 6 6 6 6 

Employee Name이 비어있지 않거나(or) Abutments.Hour.Wk.1이 missing이 아닌 경우를 통하여 subset을 추출한 뒤, 이를 hrdta2라는 변수로 만들었습니다. 이후 요약을 해보니 여전히 209개의 변수 중에 210개의 레벨이 있다는 걸 확인했습니다. 데이터 중 일부가 전체 공란이 들어 있는 것 같습니다. 이 부분은 추후에 처리하도록 하겠습니다.

In [269]:
str(hrdta[[3]])
str(hrdta[[4]])

'data.frame':	22 obs. of  14 variables:
 $ Employment.Source: Factor w/ 22 levels "Billboard","Careerbuilder",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ January          : int  520 410 0 0 0 0 0 0 640 500 ...
 $ February         : int  520 410 0 5129 0 0 0 0 640 500 ...
 $ March            : int  520 410 0 0 0 0 0 0 640 500 ...
 $ April            : int  520 820 0 0 0 0 0 0 640 440 ...
 $ May              : int  0 820 0 0 0 0 0 0 640 500 ...
 $ June             : int  0 410 0 0 0 0 0 0 640 500 ...
 $ July             : int  612 410 0 0 0 0 0 0 640 440 ...
 $ August           : int  612 820 0 4892 0 0 0 0 1300 500 ...
 $ September        : int  729 820 0 0 0 0 0 0 1300 440 ...
 $ October          : int  749 1230 0 0 0 0 0 0 1300 440 ...
 $ November         : int  910 820 0 0 0 0 0 0 1300 500 ...
 $ December         : int  500 410 0 0 0 0 0 0 1300 500 ...
 $ Total            : int  6192 7790 0 10021 0 0 0 0 10980 5760 ...
'data.frame':	12 obs. of  7 variables:
 $ Position  : Factor w/ 12 levels "Ac

다음으로 hrdta[[3]]과 hrdta[[4]]에 대해 확인을 하였습니다. 실질적인 데이터 셋이라기 보다는 부수적인 정보를 담고 있는 데이터 셋이기 때문에 별도로 전처리를 해야 할 것은 안 보입니다.

In [270]:
dir.create("use")
setwd("use")

write.csv(hrdta1,'hrdta1.csv')
write.csv(hrdta2,'hrdta2.csv')